django/db/models/window.py

```diff
@@ -0,0 +1,203 @@
+            raise FieldError('Window function expressions must be aggregate functions')
+
+    def contains_aggregate(self):
+        # although the expression contains an aggregate function, the ORM
+        # handles such expressions in ways that may not always be ideal for an
+        # OVER-clause
+        return False
+
+    def _resolve_output_field(self):
+        if self._output_field is None:
+            self._output_field = self.source_expression.output_field
+
+    def resolve_expression(self, query=None, allow_joins=True, reuse=None, summarize=False, for_save=False):
+        c = self.copy()
+        c.source_expression = c.source_expression.resolve_expression(query, allow_joins, reuse, summarize, for_save)
+        c.is_summary = summarize
+        c.for_save = for_save
+
+        if c.partition_by:
+            c.partition_by = c.partition_by.resolve_expression(query, allow_joins, reuse, summarize, for_save)
+        if c.order_by:
+            c.order_by = c.order_by.resolve_expression(query, allow_joins, reuse, summarize, for_save)
+        if c.frame:
+            c.frame = c.frame.resolve_expression(query, allow_joins, reuse, for_save)
+        return c
+
+    def as_sql(self, compiler, connection, function=None, template=None):
+        connection.ops.check_expression_support(self)
+        expr_sql, params = compiler.compile(self.source_expression)
+        window_sql, window_params = [], []
+
+        if self.partition_by is not None:
+            sql_expr, sql_params = self.partition_by.as_sql(
+                compiler=compiler, connection=connection, template='PARTITION BY %(expressions)s')
+            window_sql.extend(sql_expr)
+            window_params.extend(sql_params)
+
+        if self.order_by is not None:
+            window_sql.append(' ORDER BY ')
+            order_sql, order_params = compiler.compile(self.order_by)
+            window_sql.extend(''.join(order_sql))
+            window_params.extend(order_params)
+
+        if self.frame:
+            frame_sql, frame_params = compiler.compile(self.frame)
+            window_sql.extend(' ' + frame_sql)
+            window_params.extend(frame_params)
+
+        params.extend(window_params)
+        template = template or self.template
+
+        return template % {
+            'expression': expr_sql,
+            'window': ''.join(window_sql).strip()
+        }, params
+
+    def __str__(self):
+        ordering = partition = ''
+        if self.order_by is not None:
+            ordering = 'ORDER BY ' + str(self.order_by)
+
+        if self.partition_by is not None:
+            partition = 'PARTITION BY ' + str(self.partition_by)
+
+        return '{} OVER ({}{}{})'.format(
+            str(self.source_expression),
+            str(partition),
+            str(ordering),
+            str(self.frame or ''),
+        )
+
+    def __repr__(self):
+        return '<%s: %s>' % (self.__class__.__name__, self)
+
+    def copy(self):
```

Comment:
You can probably drop this copy if `get/set_source_expressions` is implemented instead